torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+  // executor for df, including code caches
+  GraphExecutor grad_executor;
+};
+
+} // anonymous namespace
+
+// a Graph can be created via tracing, or via a language-based frontend
+// GraphExecutor runs it. It can run the same graph on many different sizes
+// and different requires_grad states, and handles specializations for each situation.
+// GraphExecutor is completely unaware of tracing or module parameters to keep the
+// tracing concerns separated.
+struct GraphExecutorImpl {
+
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize, bool symbolically_differentiable)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(symbolically_differentiable) {}
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(isDifferentiable(*this->graph)) {}
+
+  // entry point where execution begins
+  variable_tensor_list run(variable_tensor_list inputs) {
+    // this is the fallback pathway, when we cannot differentiate
+    if(!optimize || (!symbolically_differentiable && gradientIsPossible(inputs))) {
+      auto & fb = getOrCreateAutogradFallback();
+      InterpreterState state(fb);
+      tensor_list outputs;
+      state.runOneStage(std::move(inputs), outputs);
+      // note: we never unwrapped inputs, because we want autograd to record the trace
+      return outputs;
+    }
+
+    // either we can symbolically differentiate, or we do not need a gradient.
+    // go down the route where we treat the inputs as tensors
+    // and fully optimize
+    auto & implementation = getOrCompile(inputs);
+    return implementation.run(std::move(inputs));
+  }
+
+private:
+
+  static bool gradientIsPossible(const variable_tensor_list & inputs) {
+    if (!autograd::GradMode::is_enabled()) {
+      return false;
+    }
+    for (const auto & tensor : inputs) {
+      if(tensor.defined() && Variable(tensor).requires_grad())
+        return true;
+    }
+    return false;
+  }
+  static bool isDifferentiable(Graph & g) {
+    for(auto n : g.nodes()) {
+      if(!jit::isDifferentiable(n))
+        return false;
+    }
+    return true;
+  }
+
+  static std::shared_ptr<Graph> copy(Graph & g) {
+    auto new_g = std::make_shared<Graph>();
+    std::unordered_map<Value*, Value*> value_map;
+    for(auto input : g.inputs()) {
+      value_map[input] = new_g->addInput()->copyMetadata(input);
+    }
+    for(auto node : g.nodes()) {
+      auto new_node = new_g->appendNode(new_g->createClone(node, [&](Value* v) {
+        return value_map.at(v);
+      }));
+      for(size_t i = 0; i < node->outputs().size(); ++i) {
+        value_map[node->outputs()[i]] = new_node->outputs()[i];
+        new_node->outputs()[i]->copyMetadata(node->outputs()[i]);
+      }
+    }
+    for(auto output : g.outputs()) {
+      new_g->registerOutput(value_map.at(output));
+    }
+    return new_g;
```

Comment:
How about moving this to `ir.cpp`? It sounds like a useful thing in general, and I'd rather avoid rewriting it elsewhere (because we'll forget that it's already here)